src/freertos_drivers/esp32/Esp32WS2812.cxx

```diff
@@ -0,0 +1,311 @@
+            index, red, green, blue);
+        // TODO: Add conversion for color orders other than GRB
+        ledDataBuffer_[(index * 3) + 0] = green & 0xFF;
+        ledDataBuffer_[(index * 3) + 1] = red & 0xFF;
+        ledDataBuffer_[(index * 3) + 2] = blue & 0xFF;
+        if (ledCount_ > 1)
+        {
+            sem_.post();
+        }
+        else if (update)
+        {
+            LOG(VERBOSE, "ESP-WS2812(%d): Updating LEDs", channel_);
+            ESP_ERROR_CHECK(
+                rmt_write_sample(
+                    channel_, ledDataBuffer_, ledBufferSize_, true));
+            ESP_ERROR_CHECK(
+                rmt_wait_tx_done(channel_, pdMS_TO_TICKS(timeout_ms)));
+            LOG(VERBOSE, "ESP-WS2812(%d): LEDs updated", channel_);
+        }
+    }
+}
+
+void Esp32WS2812::get_led_color(const size_t index, uint8_t *red,
+                                uint8_t *green, uint8_t *blue)
+{
+    HASSERT(index < ledCount_);
+
+    // TODO: Add conversion for color orders other than GRB
+    *green = ledDataBuffer_[(index * 3) + 0];
+    *red   = ledDataBuffer_[(index * 3) + 1];
+    *blue  = ledDataBuffer_[(index * 3) + 2];
+}
+
+void Esp32WS2812::clear(const uint8_t red, const uint8_t green,
+                        const uint8_t blue, const uint32_t timeout_ms)
+{
+    // if the color values are all zero we can set the buffer data directly.
+    if (red == 0 && green == 0 && blue == 0)
+    {
+        LOG(VERBOSE, "ESP-WS2812(%d): Turning off ALL LEDs", channel_);
+        memset(ledDataBuffer_, 0, ledBufferSize_);
+    }
+    else
+    {
+        LOG(VERBOSE,
+            "ESP-WS2812(%d): Clearing all LEDs to (r:%d,g:%d,b:%d)", channel_,
+            red, green, blue);
+        // iterate through the LEDs and set each individually.
+        for(size_t index = 0; index < ledCount_; index++)
+        {
+            // TODO: Add conversion for color orders other than GRB
+            ledDataBuffer_[(index * 3) + 0] = green  & 0xFF;
+            ledDataBuffer_[(index * 3) + 1] = red & 0xFF;
+            ledDataBuffer_[(index * 3) + 2] = blue & 0xFF;
+        }
+    }
+    // if we have more than one LED notify the background update task.
+    if (ledCount_ > 1)
+    {
+        sem_.post();
+    }
+    else
+    {
+        LOG(VERBOSE, "ESP-WS2812(%d): Updating LEDs", channel_);
+        ESP_ERROR_CHECK(
+            rmt_write_sample(channel_, ledDataBuffer_, ledBufferSize_, true));
+        ESP_ERROR_CHECK(rmt_wait_tx_done(channel_, pdMS_TO_TICKS(timeout_ms)));
+        LOG(VERBOSE, "ESP-WS2812(%d): LEDs updated", channel_);
+    }
+}
+
+void *Esp32WS2812::entry()
+{
+    LOG(VERBOSE, "ESP-WS2812(%d): Update thread starting", channel_);
+    for ( ; /* forever */ ; )
+    {
+        LOG(VERBOSE, "ESP-WS2812(%d): Waiting for update request", channel_);
+        sem_.wait();
+        ESP_ERROR_CHECK(
+            rmt_write_sample(channel_, ledDataBuffer_, ledBufferSize_, true));
```

Comment:
this could take a long time. During this time your semaphone could accumulate a lot of posts. It might make sense to eat up all counts from the semaphore after wait has returned but before this runs. This can avoid the background thread running all the time when there are no updates.

Reply:
Is there an example of that?